### PR TITLE
Cagra template instantiations

### PR DESCRIPTION
**Pull request by tfeher**
Cagra was introduced header only in https://github.com/rapidsai/raft/pull/1375. This PR adds a precompiled single- and multi-cta search kernels to libraft.so. 

The single- and multi-cta search kernels were moved to separate header files to make it easier to specify extern template instantiations for these. 

The macros for dispatching the kernels were replaced by functions. We define explicit instantiations for the top level dispatch functions. (This is in contrast to #1428 where the kernels themselves were instantiated, which resulted in a large number of parameter combinations that had to be explicitly spelled out.)

This PR fixes https://github.com/rapidsai/raft/issues/1443.

### CODE REVIEW

**Comment by tfeher** [comment]
Note for reviewers, it is best to review this PR commit by commit:
- Move CAGRA search kernels to separate header files https://github.com/rapidsai/raft/pull/1650/commits/87c501a0464c75d61c0974213496cb6f777bb7c6
- Replace CAGRA search kernel dispatch macros with functions https://github.com/rapidsai/raft/pull/1650/commits/700112fd1a2de109df5918748a7f3307bb0fdbd4
- Add explicit template instantiations for CAGRA https://github.com/rapidsai/raft/pull/1650/commits/d3300bd1d1066bbfa0f2b381f257399f3aff6976

File: cpp/test/CMakeLists.txt [hunk 1]
```diff
@@ -319,6 +319,14 @@ if(BUILD_TESTS)
     test/neighbors/ann_cagra/test_int8_t_uint32_t.cu
     test/neighbors/ann_cagra/test_uint8_t_uint32_t.cu
     test/neighbors/ann_cagra/test_float_int64_t.cu
+    src/neighbors/detail/cagra/search_multi_cta_float_uint64_dim128_t8.cu
```
Comment (tfeher) [inline cpp/test/CMakeLists.txt:322]: Note: to libraft.so we add precompiled cagra kernels for 32-bit index type only. We have test for 64-bit as well, therefore we add extra object files for the tests here.

**Review by divyegala — COMMENTED** [review]
The PR looks good to me! Just want to make sure the compile time decrease/binary size increase is roughly equivalent to #1428 before I approve this since we add another templated function in this PR.

**Review by cjnolet — APPROVED** [review]
I think this is good enough in its current state to approve for 23.08. I do think we should continue to work on the binary size and build times, but that's also a longer-term need.

File: cpp/test/neighbors/ann_cagra/test_float_int64_t.cu [hunk 2]
```diff
@@ -16,8 +16,8 @@
 
 #include <gtest/gtest.h>
 
-#undef RAFT_EXPLICIT_INSTANTIATE_ONLY
 #include "../ann_cagra.cuh"
+#include "search_kernel_uint64_t.cuh"
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/test/neighbors/ann_cagra/test_float_int64_t.cu:20]: Yes, I can make the change in a follow up PR. Under the hood IdxT is converted to unsigned type, so it is only a matter of testing the outer wrapper layers.

**Pull request merged**